### PR TITLE
Use k2 pruned transducer loss to train conformer-transducer model

### PR DESCRIPTION
**Pull request by pkufool**
Almost the same as https://github.com/k2-fsa/icefall/pull/170, except that it adds some linear transforms in `decoder.py` and `joiner.py` to make it work for any vocab-size. The pruned loss requires the output of encoder/predictior has the shape of [B, T/S, vocab-size].

You can see the previous experiment results here https://github.com/k2-fsa/icefall/pull/170.

I need to re-check the result of this PR before merging it.

### CODE REVIEW

File: egs/librispeech/ASR/pruned_transducer_stateless/joiner.py [hunk 1]
```diff
@@ -0,0 +1,51 @@
...
+
+import torch
+import torch.nn as nn
+
+
+class Joiner(nn.Module):
+    def __init__(self, input_dim: int, inner_dim: int, output_dim: int):
+        super().__init__()
+
+        self.inner_linear = nn.Linear(input_dim, inner_dim)
```
Comment (pkufool) [inline egs/librispeech/ASR/pruned_transducer_stateless/joiner.py:25]: A Linear layer here, to avoid a big matrix of (vocab-size, vocab-size) if we use large vocab-size.

File: egs/librispeech/ASR/pruned_transducer_stateless/decoder.py [hunk 2]
```diff
@@ -0,0 +1,100 @@
...
+        if context_size > 1:
+            self.conv = nn.Conv1d(
+                in_channels=embedding_dim,
+                out_channels=embedding_dim,
+                kernel_size=context_size,
+                padding=0,
+                groups=embedding_dim,
+                bias=False,
+            )
+        self.output_linear = nn.Linear(embedding_dim, vocab_size)
```
Comment (pkufool) [inline egs/librispeech/ASR/pruned_transducer_stateless/decoder.py:73]: A Linear layer to transform output dim of decoder to vocab-size.

File: egs/librispeech/ASR/pruned_transducer_stateless/decode.py [hunk 3]
```diff
@@ -0,0 +1,476 @@
...
+        }
+    )
+    return params
+
+
+def get_encoder_model(params: AttributeDict) -> nn.Module:
+    # TODO: We can add an option to switch between Conformer and Transformer
+    encoder = Conformer(
+        num_features=params.feature_dim,
+        output_dim=params.vocab_size,
```
Comment (pkufool) [inline egs/librispeech/ASR/pruned_transducer_stateless/decode.py:157]: Change the output_dim of encoder to vocab_size.

**Comment by pkufool** [comment]
The best result of this PR is as follow, we only trained 30 epochs as our GPUs are busy, you can see that the added linear layers won't affect the WERs.


<div class="okr-block-clipboard" data-okr="%7B%22okrDelta%22%3A%5B%7B%22lineType%22%3A%22unsupport%22%2C%22lineOptions%22%3A%7B%7D%2C%22lineContent%22%3A%5B%5D%7D%5D%2C%22businessKey%22%3A%22lark-doc%22%7D"></div><div data-zone-id="0" data-line-index="0" style="white-space: pre;">

Settings | WERs (test-clean) | WERs (test-other) | Notes
-- | -- | -- | --
Baseline (model without linear layers) max-duration=300 | 2.89 | 7.07 | --epoch=30, --avg=12
Model with linear layers max-duration=300 | 2.85 | 6.98 | --epoch=28, --avg=15
</div>

**Comment by danpovey** [comment]
So do you think now is the right time to merge it?

**Comment by pkufool** [comment]
> So do you think now is the right time to merge it?

I think it is ready now.

**Comment by csukuangfj** [comment]
@pkufool 

Could you upload your pre-trained model to hugging face and provide a colab notebook to demonstrate how to use your pre-trained model?

**Comment by pkufool** [comment]
OK, I think it is the same as your stateless transducer, it only changes the loss part.

**Comment by desh2608** [comment]
@pkufool do you have the tensorboard for this training by any chance? I am training a pruned RNNT model on a different dataset (SPGISpeech; 5k hours) and was wondering what kind of loss values the model usually converges to?
